contributors/design-proposals/auth/external-secrets.md

```diff
@@ -0,0 +1,218 @@
+apiVersion: v1
+kind: Pod
+metadata:
+  name: test-pod
+spec:
+  containers:
+  - image: registry.access.redhat.com/rhel7
+    imagePullPolicy: Never
+    name: test-pod
+    volumeMounts:
+    - mountPath: /mnt/secrets-external
+      name: secrets-external
+  volumes:
+  - flexVolume:
+      driver: example.com/secrets-fuse
+    name: secrets-external
+```
+
+Within the container of the pod, the application can then access
+`/mnt/secrets-external/HTTP/www.example.com` and the FUSE filesystem
+will see this as access to `HTTP/www.example.com` file. Via the
+`kubernetes.io/pod.uid`, it can verify (against its internal rules
+or for example via Kubernetes API against attributes of that Pod)
+whether the Pod should be allowed access to the `HTTP/www.example.com`
+secret, and then invoke tools like `custodia-cli` via
+
+```
+custodia-cli get HTTP/www.example.com
+```
+
+to get the secret from remote, external source. Alternatively, the
+Pod identity with other attributes / annotations can be passed to the
+remote provider to run the authorization on the remote side.
+
+Since the FUSE provider runs as root, it will have access to necessary
+credentials to prove its own identity to the remote provider.
+
+In production setup, we expect that remote provider to be a Custodia
+server which will make additional verification against Kubernetes API,
+checking that the on-Node identity which makes the request for
+secrets is only making the request on behalf of the Pods that are
+currently running on the Node. That way, if a Node is compromised,
+only (secrets of) Pods running on that single Node are compromised.
+The Custodia server on that separate host will then proxy the requests
+to the actual vault or secrets provider.
+
+## Implementation
+
+Implementation of FUSE filesystem using libfuse 2.9 will show example
+invocation of `custodia-cli` to get secrets from remote Custodia server.
+Accompanying flexVolume driver would then mount the FUSE filesystem
+when configured for Pod.
+
+The flexVolume's `init` has to return `"selinuxRelabel": false` to
+prevent labelling operation which is not supported on the FUSE
+filesystem.
+
+The FUSE filesystem has to be mounted with `allow_other` option to
+allow non-root uids (that the containers run as) to access the filesystem.
+
+The `kubernetes.io/pod.uid` of `mount` parameters needs to be parsed.
+
+During `getattr` and `read`, `struct fuse_context *fuse_get_context(void)`
+can be called to get pid, uid, and git of the calling process. This
+information can be used to cross-verify it against the the Pod
+identity to prevent leaking the information from the mounted
+filesystem on the host, for example during backup.
+
+The SELinux boolean `virt_sandbox_use_fusefs` has to be set to true to
+allow Kubernetes containers to access the filesystem.
+
+### Client/Server Backwards/Forwards compatibility
+
+Not affected.
+
+## Alternatives considered
+
+Container Storage Interface approach might be the long-run preferred
```

Comment:
If this proposal is "flex volume can be used to inject secrets" (which is good to document) there needs to be more discussion of different ways and tradeoffs that flex volumes can be used (you talk a bit about a daemon socket, but don't discuss injecting static tmpfs volumes or other solutions in use in the space).  You want to make the case that flex volume *should* be used for this use case until a later point which may include a new storage mechanism.  I would also call out that evolution of this use case should inform the CSI group and so that there would be a request for sig-storage to review the proposal and the case you are making so that they can understand and comment as to the direction of CSI.